teps/0092-scheduling-timeout.md

```diff
@@ -17,361 +24,474 @@ authors:
+    total: "10m" # optional, if specified, from TaskRun.StartTime to TaskRun.CompletionTime (defaults to 60m)
+  timeout: "10m" # existing deprecated field, from TaskRun.StartTime to TaskRun.CompletionTime (defaults to 60m)
+```
 
-### PipelineRun
+When set, the order of timeout application would be:
+1. At the start of the `TaskRun`, the `scheduling` and `total` timeouts counters would begin
+2. Once the `Pod` executing the `TaskRun` has been scheduled, the `scheduling` timeout would stop and the `execution` timeout would begin.
+3. The completion of the `TaskRun` must be before the `execution` and `total` timeout counters end.
 
-A new section `taskruns` would be added to
-[the existing timeout configuration in PipelineRun](https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#configuring-a-failure-timeout)
-with one new option called `scheduling`:
+Note that:
+
+* If no timeouts are set, only the default `total` timeout is used
+* Note that the sum of `scheduling` timeout and `execution` timeout must equal the `total` timeout
+* If either `scheduling` or `execution` timeout is set to 0, then total timeout must also be set to 0
+
+
+### Examples
+
+###### User has a resource-constrained environment and wants to cancel if not scheduled quickly
+
+User creates this:
 
 ```yaml
 spec:
   timeouts:
-    pipeline: "0h15m0s"
-    tasks: "0h10m0s"
-    finally: "0h5m0s"
-    taskruns: # this section would indicate that this timeout should be applied to all pipeline tasks
-      scheduling: "1m" # this is the new option
+    scheduling: "5m"
 ```
 
-When set:
-* This timeout would be applied to all `TaskRuns` created when executing the `Pipeline`
-* The order of timeout application would be:
-  1. At the start of the `Pipeline` execution, the `pipeline` and `tasks` timeout counters would begin
-  2. At the start of each `Task` being executed as a `TaskRun`, the `scheduling` timeout counter would begin,
-     considering the timeout to have started from the creation time of the `TaskRun` in etcd
-  3. Once the pod corresponding to the `TaskRun` has been scheduled, if the Pipeline Task configuration
-     [includes a timeout for the individual Task,](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#configuring-the-failure-timeout)
-     the counter for that timeout would begin
-  4. Once the `finally` Tasks in the `Pipeline` begin executing, the `tasks` timeout counter would no loner be counting,
-     and the `finally` counter would start
-  5. The `scheduling` timeout would apply to the `finally` Tasks in the same way as it is applied to the other tasks
+The spec that is calculated is this:
 
-When not set, no scheduling timeout would apply (i.e. there is no scheduling timeout by default).
+```yaml
+spec:
+  timeouts:
+    scheduling: "5m"  # User-specified value
+    execution: "55m"  # Calculated from default total timeout - scheduling timeout
+    total: "60m"  # Default value
```

Comment:
So this one is a bit weird/confusing to me. Ideally, I want the scheduling to not take more than `5m` and the whole *execution* of my Task (scheduling including) the default value, so 60m. My assumption is then that, the `TaskRun` wouldn't time-out before 60m has passed, no matter how much time the scheduling did take. This is, however, not what happens in the proposal. If the scheduling took about 10 second, then, the `TaskRun` would timeout after 56min and not 60min (because the execution is still gonna timeout after 55m, even though we had more time for the total).

Reply:
Thanks for the catch. Agreed. The execution time could range from 55 mins to 60 minutes depending on how long scheduling took. The TEP has been updated to reflect this.